framer/InputLayer.coffee

```diff
@@ -0,0 +1,151 @@
+{Events} = require "./Events"
+
+Events.EnterKey  = "EnterKey"
+Events.BackSpaceKey  = "BackSpaceKey"
+Events.InputValueChange = "InputValueChange"
+Events.InputFocus = "InputFocus"
+Events.InputBlur = "InputBlur"
+
+class exports.InputLayer extends TextLayer
+
+	constructor: (options={}) ->
+
+		_.defaults options,
+			backgroundColor: "#FFF"
+			width: 500
+			height: 100
+
+		super options
+
+		# Layer containing input element
+		@input = new Layer
+			backgroundColor: "transparent"
+			x: 30
+			width: @width - 60
+			height: @height
+			parent: @
+			name: "input"
+
+		if not @multiLine
+			@_inputElement = document.createElement("input")
+		else
+			@_inputElement = document.createElement("textarea")
+			@input.y = 25
+
+		# The id serves to differentiate multiple input elements from one another.
+		# To allow styling the placeholder colors of seperate elements.
+		@_id = Utils.round(Utils.randomNumber(0, 100))
+		@_inputElement.className = "input" + @_id
+
+		@on "change:width", =>
+			@_inputElement.style.width = "#{@input.width}px}"
+
+		# Append element
+		@input._element.appendChild(@_inputElement)
+
+		# Match TextLayer defaults and type properties
+		@_setTextProperties = =>
+			@_inputElement.style.fontFamily = @fontFamily
+			@_inputElement.style.fontSize = @fontSize
+			@_inputElement.style.lineHeight = @lineHeight
+			@_inputElement.style.fontWeight = @fontWeight
+			@_inputElement.style.color = @color
+			@_inputElement.style.outline = "none"
+			@_inputElement.style.backgroundColor = "transparent"
+			@_inputElement.style.width = "#{@width - 64}px"
+			@_inputElement.style.height = "#{@height}px"
+			@_inputElement.style.cursor = "auto"
+			@_inputElement.style.webkitAppearance = "none"
+			@_inputElement.style.resize = "none"
+
+		@_setTextProperties()
+
+		# All inherited properties
+		@_textProperties =
+			{@fontFamily, @fontSize, @lineHeight, @fontWeight, @color, @backgroundColor, @width, @height}
+
+		for property, value of @_textProperties
+
+			@on "change:#{property}", (value) =>
+				@_setTextProperties()
+				@_setPlaceholder()
+
+		# If text has been defined, use that, otherwise default to placeholder
+		@_defaultText = @text
+		@_setPlaceholder()
+
+		# Override text property setting the html
+		@html = ""
+
+		# Check if in focus
+		@_isFocused = false
+
+		# Default focus interaction
+		@_inputElement.onfocus = (e) =>
+
+			@_inputElement.style.color = "#000"
+
+			# Emit focus event
+			@emit(Events.InputFocus, event)
+
+			@_isFocused = true
+
+		# On blur event
+		@_inputElement.onblur = (e) =>
+			# Emit blur event
+			@emit(Events.InputBlur, event)
+
+		@_inputElement.onkeyup = (e) =>
+
+			# Check last character
+			lastCharacter = @_inputElement.value.substr(@_inputElement.value.length - 1)
+
+			# Exclude enter, shift, caps lock, control, alt, command etc.
+			unless e.which in [13, 16, 17, 18, 20, 91, 93]
+				@emit("change:value", @value)
+				@emit(Events.InputValueChange, @value)
```

Comment:
Ha wow, this is so much better indeed. Immediately solves the problem of other character combinations/keys I may have forgotten and whatnot. See https://github.com/koenbok/Framer/commit/c7397348e92b2cacc15e2e1ab179f69adb481b0f